sycl/doc/Assert.md

```diff
@@ -0,0 +1,269 @@
+# Assert feature
+
+**IMPORTANT**: This document is a draft.
+
+Using the standard C++ `assert` API ("assertions") is an important debugging
+technique widely used by developers. This document describes the design of
+supporting assertions within SYCL device code.
+The basic approach we chose is delivering device-side assertions as host-side
+asynchronous exceptions, which allows further extensibility, such as better
+error handling or potential recovery.
+
+As usual, device-side assertions can be disabled by defining `NDEBUG` macro at
+compile time.
+
+## Use-case example
+
+```
+using namespace sycl;
+auto ErrorHandler = [] (exception_list Exs) {
+  for (exception_ptr const& E : Exs) {
+    try {
+      std::rethrow_exception(E);
+    }
+    catch (event_error const& Ex) {
+      std::cout << “Exception - ” << Ex.what(); // assertion failed
+      std::abort();
+    }
+  }
+};
+
+void user_func(item<2> Item) {
+  assert((Item[0] % 2) && “Nil”);
+}
+
+int main() {
+  queue Q(ErrorHandler);
+  q.submit([&] (handler& CGH) {
+    CGH.parallel_for<class TheKernel>(range<2>{N, M}, [=](item<2> It) {
+      do_smth();
+      user_func(It);
+      do_smth_else();
+    });
+  });
+  Q.wait_and_throw();
+  std::cout << “One shouldn’t see this message.“;
+  return 0;
+}
+```
+
+In this use-case every work-item with even X dimension will trigger assertion
+failure. Assertion failure should be reported via asynchronous exceptions. If
+asynchronous exception handler is set the failure is reported with
+`sycl::event_error` exception. Otherwise, SYCL Runtime should trigger abort.
+Even though multiple failures of the same or different assertions can happen in
+multiple workitems, implementation is required to deliver only one. The
+assertion failure message is printed to `stderr` by SYCL Runtime.
+
+When multiple kernels are enqueued and more than one fail at assertion, at least
+single assertion should be reported.
+
+
+## User requirements
+
+From user's point of view there are the following requirements:
+
+| # | Title | Description | Importance |
+| - | ----- | ----------- | ---------- |
+| 1 | Handle assertion failure | Signal about assertion failure via SYCL asynchronous exception | Must have |
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+Implementations without enough capabilities to implement fourth requirement are
+allowed to realize the fallback approach described below, which does not
+guarantee assertion failure delivery to host, but is still useful in many
+practical cases.
+
+## Contents of `sycl::event_error`
+
+Interface of `sycl::event_error` should look like:
+```
+class event_error : public runtime_error {
```

Comment:
Note: [SYCL 2020](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:exception.class) used different approach for exceptions: instead of several sub-classes, there is now just a single `sycl::exception` class, which contains an error code describing what happened wrong

Reply:
Added extension document for assert error code